source/common/http/async_client_impl.cc

```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
       dispatcher_(dispatcher) {}
 
 AsyncClientImpl::~AsyncClientImpl() {
-  while (!active_requests_.empty()) {
-    active_requests_.front()->failDueToClientDestroy();
+  while (!active_streams_.empty()) {
+    active_streams_.front()->reset();
   }
 }
 
 AsyncClient::Request* AsyncClientImpl::send(MessagePtr&& request, AsyncClient::Callbacks& callbacks,
                                             const Optional<std::chrono::milliseconds>& timeout) {
-  std::unique_ptr<AsyncRequestImpl> new_request{
-      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout)};
+  AsyncRequestImpl* async_request =
```

Comment:
Having a raw pointer here is to avoid downcasting from `AsyncStreamImpl` to `AsyncRequestImpl` when it returns. `AsyncStreamImpl` is not a subclass of `AsyncClient::Request`.